crates/sparrow-backend/src/pipeline_schedule.rs

```diff
@@ -0,0 +1,163 @@
+use sparrow_core::debug_println;
+use sparrow_physical::{Pipeline, Step, StepKind};
+
+const DEBUG_SCHEDULING: bool = false;
+
+/// Determine the pipeline each step should be part of.
+///
+/// This is computed in a bottom-up manner:
+///
+/// 1. Each leaf (input) is the start of a separate pipeline.
+/// 2. Certain step kinds are "pipeline breaking", which means
+///    they start a new pipeline. For instance, `with_key`.
+/// 3. Any other operation with a single input is part of the
+///    the same pipeline as the input.
+/// 4. Any other operation is a separate pipeline.
+pub fn pipeline_schedule(steps: &[Step]) -> Vec<Pipeline> {
+    // Compute how many times each step is referenced.
+    // Any step referenced multiple times must end a pipeline.
+    let mut references = vec![0; steps.len()];
+    for step in steps {
+        for input in &step.inputs {
+            references[input.0] += 1;
+        }
+    }
+
+    // Then compute the assignments for each step.
+    let mut assignments = Vec::with_capacity(steps.len());
+
+    let mut pipelines = Vec::new();
+    for (index, step) in steps.iter().enumerate() {
+        let break_pipeline = if step.inputs.len() != 1 {
```

Comment:
This was helpful for debugging, but makes the code a lot longer. Really, the criteria is:

```
step.inputs.len() != 1 || references[step.inputs[0.0] > 1 || is_pipeline_breaker(&step.kind)
```

My thought was to keep the longer form around, since it *is* useful to explain why a pipeline was broken... but I could be convinced to drop this for simplicity.

Reply:
I imagine we'll want the debugging. Could refactor into a `breaks_pipeline()` method to shorten this method.